Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -43,13 +43,28 @@ std::map<std::string, FieldValue> DecodeObject(pb_istream_t* stream);
 
 /**
  * Docs TODO(rsgowman). But currently, this just wraps the underlying nanopb
- * pb_ostream_t. Eventually, this might use static factory methods to create the
- * underlying pb_ostream_t rather than directly passing it in.
+ * pb_ostream_t.
  */
 // TODO(rsgowman): Encode* -> Write*
 class Writer {
  public:
-  explicit Writer(pb_ostream_t* stream) : stream_(stream) {
+  /**
+   * Creates an output stream that writes to the specified vector. Note that
+   * this vector pointer must remain valid for the lifetime of this Writer.
```

Comment:
Can `Writer` create the vector and provide an accessor like `const uint8_t* data() const;`?

Reply:
Ah, sorry, it should be `Serializer`, not `Writer`, that owns the vector. However, I assumed that `Serializer` lives as long as the calling code keeps it alive (and thus can encode several top-level messages in a row and then move out the underlying vector). If that's not the case, then it would be more trouble than it's worth.

Reply:
Ah, that makes more sense (and I think eliminates difficulties 1 and 3.) I think 2 is still a valid concern though... OTOH, maybe we won't end up using that. (Certainly we can't without adjusting the api a bit.)

Aside: We might end up adjusting the API to, rather than taking a vector*, instead take in a Writer. That would allow the user to create Writer::FromBuffer(), ::FromNetwork(), ::FromDisk(), etc as required and pass it in to the serializer class. (Well, FromBuffer -> Wrap, FromEtc -> ?)

For now, I've added a TODO to the header:
```
  // TODO(rsgowman): If we never support any output except to a vector, it may
  // make sense to have Serializer own the vector and provide an accessor rather
  // than asking the user to create it first.
```

Reply:
Thanks!

For 2, I'm not exactly sure how it would look like, but perhaps it's a job for several Writer classes? Or maybe abstract away the output destination into its own class?

Reply:
I'm not sure case (2) is anything to worry about. Writing to gRPC requires writing to buffers that are then handed over to it. We will never write directly to some network socket or something like that. Similarly, writing to leveldb or sqlite requires the construction of a complete value to store in a buffer so we'll never write directly to disk.

The type of the buffer we use is likely to change, and in that regard it's a good thing to pass `Writer*` around as our internal type, but it's not worth any effort to try to abstract away from writing to a buffer.

Reply:
I don't remember how calling code looks like, but if it needs to own the vector after `Serializer` is destroyed, maybe provide a way to take ownership of the vector?

```cpp
class Serializer { /*...*/
  std::vector<unit8_t> DisownBuffer(); // Probably not under that name
};
```

Reply:
In the other clients the Serializer is a singleton, with possible polymorphic behavior based on variations in the protocol (see typescript where we support serialization of v1beta1 to JSON suitable for construction of protocol buffers and JSON suitable for sending via API-JSON over webchannel--these differ in how e.g. dates are represented).

I don't see any great incentive to make this serializer any different, and I have no problem passing in the buffer.

Meanwhile, I think exactly how this is to be handled depends a lot on how we end up handling higher-level constructs like Documents that contain Values so I think it's premature to spend much effort changing this ownership around. Let's get some working subset of Document document serialization working before worrying too much about this contract.

Reply:
So to sum up (and put words in everyone's mouths) it sounds like we're largely ok committing this as is but will revisit this issue when we actually start using this to ensure the api still makes sense. I'll leave the TODO in for now to remind us.